### DISCUSSION TITLE
Read text in correct order from pdf?

### DISCUSSION THREAD

**Discussion by Rapid1898-code** [jsvine/pdfplumber#850] [Ask for help with specific PDFs]
Hello - i would like to read the full text of a pdf (from lines and tables) in the correct order -

Currently i am using the following code for this for eg. the attached pdf -
[partly.pdf](https://github.com/jsvine/pdfplumber/files/11131608/partly.pdf)

```
import pdfplumber
import os
import sys

if __name__ == '__main__':
  path = os.path.abspath(os.path.dirname(sys.argv[0])) 
  fn = os.path.join(path, "partly.pdf")
  table_settings = {
      "vertical_strategy": "lines",
      "horizontal_strategy": "lines"
  }
  pdf = pdfplumber.open(fn)    
  
  workTable= []
  workLines = []
  for idxPage, page in enumerate(pdf.pages, start=1):                         
    print(f"Reading tables for page {idxPage}")
    # read the content of the tables 
    erg=page.extract_tables(table_settings)
    for idxTable, table in enumerate(erg, start=1):
      for row in table:
        if row[0] in ["", None]:
          continue
        wRow = [idxPage, row[0]]
        worker = row[1:]
        worker = [x for x in worker if x not in ["",None]]
        worker = " ".join(worker)
        wRow.append(worker)
        workTable.append(wRow)        
    # read the text   line by line from the tables and try to parse
    workLines.extend(page.extract_text().splitlines())
    
  for e in workTable:
    print(e)
  for l in workLines:
    print(l)
```

Generally this works fine but now i want to bring all the output in order / correct position of the document.
So first the text - then the table content - next text - next table content and so on.
How can i do this eg. for the attached document?

**Comment by jsvine** [comment]
The particular strategy will depend on the particular way that the tables are integrated into the rest of the layout, but two things that might help:

- The `table.bbox` attribute, which will give you the table's bounding box
- Using `page.extract_words(...)`, which will give you objects that contain the text and positioning of each word on the page, in sorted order, and which you can compare to the positions of the tables
